File: backend/src/shared_lib/entities/finding.py
from typing import TypedDict

from pydantic import BaseModel

FindingID = str


class Finding(BaseModel):
    id: FindingID
    status_detail: str | None = None
    risk_details: str | None = None


class FindingMetaData(BaseModel):
    event_code: str | None = None


class FindingResource(BaseModel):
    uid: str | None = None
    name: str | None = None
    type: str | None = None
    region: str | None = None


class FindingRemediation(BaseModel):
    desc: str | None = None
    references: list[str] | None = None


class FindingExtra(Finding):
    metadata: FindingMetaData | None = None
    status_code: str | None = None
    severity: str | None = None
    resources: list[FindingResource] | None = None
    remediation: FindingRemediation | None = None
    hidden: bool = False


class FindingDto(BaseModel):
    hidden: bool | None = None


class FindingRule(TypedDict):
    pillar: str
    question: str
    best_practice: str


FilteringRules = dict[str, list[FindingRule]]

File: libs/backend/models/src/lib/Finding/Finding.ts
import { BestPractice } from '../models';

type AggregationFieldSelection<T> = {
  [K in keyof T]?: T[K] extends Array<infer U>
    ? AggregationFieldSelection<U>
    : T[K] extends object
      ? AggregationFieldSelection<T[K]>
      : true;
};

type AggregationFieldResult<T> = {
  [K in keyof T]: NonNullable<T[K]> extends true
    ? Record<string, number>
    : NonNullable<T[K]> extends object
      ? AggregationFieldResult<NonNullable<T[K]>>
      : never;
};

export type FindingAggregationFields = AggregationFieldSelection<Finding>;
export type FindingAggregationResult<TFields extends FindingAggregationFields> =
  AggregationFieldResult<TFields>;

export interface Finding extends ScanFinding {
  hidden: boolean;
  isAIAssociated: boolean;
  bestPractices?: BestPractice[];
  comments?: FindingComment[];
}

export interface ScanFinding {
  id: string;
  eventCode?: string;
  remediation?: FindingRemediation;
  resources: FindingResource[];
  riskDetails: string;
  severity: SeverityType;
  statusCode: string;
  statusDetail: string;
}

export interface FindingRemediation {
  desc: string;
  references: string[];
}

export interface FindingResource {
  name?: string;
  region?: string;
  type?: string;
  uid?: string;
}

export enum SeverityType {
  Unknown = 'Unknown',
  Informational = 'Informational',
  Low = 'Low',
  Medium = 'Medium',
  High = 'High',
  Critical = 'Critical',
  Fatal = 'Fatal',
  Other = 'Other',
}

export interface FindingComment {
  id: string;
  authorId: string;
  text: string;
  createdAt: Date;
}

export interface FindingCommentBody {
  text?: string;
}

export interface FindingBody {
  hidden?: boolean;
}
